pilosa/client.py

```diff
@@ -200,30 +201,37 @@ def _import_bits(self, index_name, frame_name, slice, bits):
             client._import_node(_ImportRequest(index_name, frame_name, slice, bits))
 
     def _fetch_fragment_nodes(self, index_name, slice):
-        uri = "%s/fragment/nodes?slice=%d&index=%s" % (self.__get_address(), slice, index_name)
-        response = self.__http_request("GET", uri, client_response=Client.__ERROR_CHECKED_RESPONSE)
+        path = "/fragment/nodes?slice=%d&index=%s" % (slice, index_name)
+        response = self.__http_request("GET", path, client_response=Client.__ERROR_CHECKED_RESPONSE)
         content = response.data.decode('utf-8')
         return json.loads(content)
 
     def _import_node(self, import_request):
         data = import_request.to_protobuf()
-        uri = "%s/import" % self.__get_address()
-        self.__http_request("POST", uri, data=data)
+        self.__http_request("POST", "/import", data=data, client_response=Client.__RAW_RESPONSE)
 
     def __patch_index_time_quantum(self, index):
-        uri = "%s/index/%s/time-quantum" % (self.__get_address(), index.name)
+        path = "/index/%s/time-quantum" % index.name
         data = '{\"timeQuantum\":\"%s\"}"' % str(index.time_quantum)
-        self.__http_request("PATCH", uri, data=data)
+        self.__http_request("PATCH", path, data=data)
 
-    def __http_request(self, method, uri, data=None, client_response=0):
+    def __http_request(self, method, path, data=None, client_response=0):
         if not self.__client:
             self.__connect()
-        try:
-            response = self.__client.request(method, uri, body=data)
-        except urllib3.exceptions.MaxRetryError as e:
-            self.cluster.remove_host(self.__current_host)
-            self.__current_host = None
-            raise PilosaError(str(e))
+        # try at most 10 non-failed hosts; protect against broken cluster.remove_host
+        for _ in range(_MAX_HOSTS):
+            uri = "%s%s" % (self.__get_address(), path)
+            try:
+                _LOGGER.debug("Request: %s %s %s", method, uri,
+                              "[binary]" if client_response == Client.__RAW_RESPONSE else data)
+                response = self.__client.request(method, uri, body=data)
+                break
+            except urllib3.exceptions.MaxRetryError as e:
+                self.cluster.remove_host(self.__current_host)
```

Comment:
I've modified the Cluster class, so `_reset` is automatic. The Client copies the cluster on construction, so client intervention is not required nor allowed. If all hosts in the cluster are blacklisted, an exception is raised and the user should handle that (maybe just calling the request again).

Changing internal state in the client between requests should not be an issue; if a host is blacklisted after a request, it should still be blacklisted for other requests.

The code in the Client which makes use of the cluster should be pretty generic; the user may write a Cluster class which is more or less complex than the one we have.